Index: tesseract-2.04/viewer/svutil.cpp
===================================================================
--- tesseract-2.04.orig/viewer/svutil.cpp	2015-08-05 15:39:10.825224457 +0200
+++ tesseract-2.04/viewer/svutil.cpp	2015-08-05 15:42:17.257224944 +0200
@@ -28,6 +28,7 @@
 #include <windows.h>
 #include <winsock.h>
 #else
+#include <stdio.h>
 #include <arpa/inet.h>
 #include <netinet/in.h>
 #include <pthread.h>
@@ -38,6 +39,7 @@
 #include <netdb.h>
 #include <sys/socket.h>
 #ifdef __linux__
+#include <unistd.h>
 #include <sys/prctl.h>
 #endif
 #endif
